velox/exec/Expand.cpp

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+#include "velox/exec/Expand.h"
+
+namespace facebook::velox::exec {
+
+Expand::Expand(
+    int32_t operatorId,
+    DriverCtx* driverCtx,
+    const std::shared_ptr<const core::ExpandNode>& expandNode)
+    : Operator(
+          driverCtx,
+          expandNode->outputType(),
+          operatorId,
+          expandNode->id(),
+          "Expand") {
+  const auto& inputType = expandNode->sources()[0]->outputType();
+  auto numRows = expandNode->projections().size();
+  rowProjections_.reserve(numRows);
+  constantProjections_.reserve(numRows);
+  auto numColumns = expandNode->names().size();
+  for (const auto& rowProjections : expandNode->projections()) {
+    std::vector<column_index_t> rowProjection;
+    rowProjection.reserve(numColumns);
+    std::vector<ConstantTypedExprPtr> constantProjection;
+    constantProjection.reserve(numColumns);
+    for (const auto& columnProjection : rowProjections) {
+      if (auto field =
+              std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(
+                  columnProjection)) {
+        rowProjection.push_back(inputType->getChildIdx(field->name()));
+        constantProjection.push_back(nullptr);
+      } else if (
+          auto constant =
+              std::dynamic_pointer_cast<const core::ConstantTypedExpr>(
+                  columnProjection)) {
+        rowProjection.push_back(kUnMapedProject);
+        constantProjection.push_back(constant);
+      } else {
+        VELOX_USER_FAIL(
+            "Expand operator doesn't support this expression. Only column references and constants are supported. {}",
+            columnProjection->toString())
+      }
+    }
+
+    rowProjections_.emplace_back(std::move(rowProjection));
+    constantProjections_.emplace_back(std::move(constantProjection));
+  }
+}
+
+bool Expand::needsInput() const {
+  return !noMoreInput_ && input_ == nullptr;
+}
+
+void Expand::addInput(RowVectorPtr input) {
+  // Load Lazy vectors.
+  for (auto& child : input->children()) {
+    child->loadedVector();
+  }
+
+  input_ = std::move(input);
+}
+
+RowVectorPtr Expand::getOutput() {
+  if (!input_) {
+    return nullptr;
+  }
+
+  auto numInput = input_->size();
+
+  std::vector<VectorPtr> outputColumns(outputType_->size());
+
+  const auto& rowProjection = rowProjections_[rowIndex_];
+  const auto& constantProjection = constantProjections_[rowIndex_];
+  auto numColumns = rowProjection.size();
+
+  for (auto i = 0; i < numColumns; ++i) {
+    if (rowProjection[i] == kUnMapedProject) {
+      const auto& constantExpr = constantProjection[i];
+      if (constantExpr->value().isNull()) {
+        // Add null column.
+        outputColumns[i] = BaseVector::createNullConstant(
+            outputType_->childAt(i), numInput, pool());
+      } else {
+        // Add constant column.
+        outputColumns[i] = BaseVector::createConstant(
+            constantExpr->type(), constantExpr->value(), numInput, pool());
```

Comment:
constantExpr->value() won't work if constant is of complex type (array, map, struct)

Use constantExpr->toConstantVector(pool()) instead.